File: implementations/rust/src/id57_short.rs
use crate::errors::B57Error;
use crate::h57::HashFunction;
use crate::id57::{
    id57_generate, id57_is_canonical, id57_is_valid, ID57Length,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum ID57ShortLength {
    Default = 0,
    Len23 = 23,
    Len29 = 29,
    Len32 = 32,
    Len47 = 47,
    Len70 = 70,
}

pub fn id57_short_generate(
    input: &[u8],
    hash_fn: Option<HashFunction>,
    length: ID57ShortLength,
) -> Result<String, B57Error> {
    let effective = resolve_id57_short_length(length)?;
    id57_generate(input, hash_fn, effective)
}

pub fn id57_short_generate_default(input: &[u8]) -> Result<String, B57Error> {
    id57_short_generate(input, Some(HashFunction::Blake3), ID57ShortLength::Default)
}

pub fn id57_short_verify(
    input: &[u8],
    hash_fn: Option<HashFunction>,
    id57_string: &str,
    length: ID57ShortLength,
) -> bool {
    id57_short_generate(input, hash_fn, length)
        .map(|expected| expected == id57_string)
        .unwrap_or(false)
}

pub fn id57_short_verify_default(input: &[u8], id57_string: &str) -> bool {
    id57_short_verify(
        input,
        Some(HashFunction::Blake3),
        id57_string,
        ID57ShortLength::Default,
    )
}

pub fn id57_short_is_valid(id57_string: &str) -> bool {
    id57_is_valid(id57_string)
}

pub fn id57_short_is_canonical(id57_string: &str) -> bool {
    id57_is_canonical(id57_string)
}

pub fn resolve_id57_short_length(length: ID57ShortLength) -> Result<ID57Length, B57Error> {
    match length {
        ID57ShortLength::Default => Ok(ID57Length::Len47),
        ID57ShortLength::Len23 => Ok(ID57Length::Len23),
        ID57ShortLength::Len29 => Ok(ID57Length::Len29),
        ID57ShortLength::Len32 => Ok(ID57Length::Len32),
        ID57ShortLength::Len47 => Ok(ID57Length::Len47),
        ID57ShortLength::Len70 => Ok(ID57Length::Len70),
    }
}
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_and_verify() {
        let s = id57_short_generate_default(b"abc").unwrap();
        assert!(id57_short_verify_default(b"abc", &s));
    }

    #[test]
    fn invalid_len_rejected_by_surface() {
        let s = id57_short_generate(b"abc", Some(HashFunction::Blake3), ID57ShortLength::Len23).unwrap();
        assert!(id57_short_is_valid(&s));
        assert!(id57_short_is_canonical(&s));
    }
}

File: implementations/ts/id57_short.ts
import {
  ID57Length, id57Generate, id57GenerateDefault,
  id57IsCanonical, id57IsValid, id57Verify, id57VerifyDefault
} from './id57.js';
import { newInvalidLengthEnumError } from './errors.js';

type ByteInput = Uint8Array | Buffer | ArrayBuffer;

export const ID57ShortLength = Object.freeze({
  DEFAULT: 0, LEN_23: 23, LEN_29: 29, LEN_32: 32, LEN_47: 47, LEN_70: 70
});

const id57ShortToID57Length: Map<number, number> = new Map([
  [ID57ShortLength.LEN_23, ID57Length.LEN_23],
  [ID57ShortLength.LEN_29, ID57Length.LEN_29],
  [ID57ShortLength.LEN_32, ID57Length.LEN_32],
  [ID57ShortLength.LEN_47, ID57Length.LEN_47],
  [ID57ShortLength.LEN_70, ID57Length.LEN_70]
]);

export function resolveID57ShortLength(length: number) {
  if (length === ID57ShortLength.DEFAULT) return ID57Length.LEN_47;
  const effective = id57ShortToID57Length.get(length);
  if (!effective) throw newInvalidLengthEnumError(length);
  return effective;
}

export function id57ShortGenerate(input: ByteInput, length: number = ID57ShortLength.DEFAULT) {
  const effective = resolveID57ShortLength(length);
  return id57Generate(input, effective);
}

export function id57ShortGenerateDefault(input: ByteInput) {
  return id57ShortGenerate(input, ID57ShortLength.DEFAULT);
}

export function id57ShortVerify(input: ByteInput, id57String: string, length: number = ID57ShortLength.DEFAULT) {
  try {
    return id57ShortGenerate(input, length) === id57String;
  } catch {
    return false;
  }
}

export function id57ShortVerifyDefault(input: ByteInput, id57String: string) {
  return id57ShortVerify(input, id57String, ID57ShortLength.DEFAULT);
}

export function id57ShortIsValid(id57String: string) {
  return id57IsValid(id57String);
}

export function id57ShortIsCanonical(id57String: string) {
  return id57IsCanonical(id57String);
}

export { id57GenerateDefault, id57Verify, id57VerifyDefault };
